source/common/shared_pool/shared_pool.h

```diff
@@ -0,0 +1,97 @@
+ *
+ * Note:  ObjectSharedPool needs to be created in the main thread,
+ * all the member methods can only be called in the main thread,
+ * it does not have the ownership of object stored internally, the internal storage is weak_ptr,
+ * when the internal storage object destructor executes the custom deleter to remove its own
+ * weak_ptr from the ObjectSharedPool.
+ */
+template <typename T, typename HashFunc = std::hash<T>,
+          class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
+class ObjectSharedPool : public Singleton::Instance,
+                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
+                         NonCopyable {
+public:
+  ObjectSharedPool(Event::Dispatcher& dispatcher)
+      : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
+  ~ObjectSharedPool() = default;
+
+  void deleteObject(const size_t hash_key) {
+    if (std::this_thread::get_id() == thread_id_) {
+      // There may be new inserts with the same hash value before deleting the old element,
+      // so there is no need to delete it at this time.
+      if (object_pool_.find(hash_key) != object_pool_.end() &&
+          object_pool_[hash_key].use_count() == 0) {
+        object_pool_.erase(hash_key);
+      }
+    } else {
+      // Most of the time, the object's destructor occurs in the main thread, but with some
+      // exceptions, it is destructed in the worker thread. In order to keep the object_pool_ thread
+      // safe, the deleteObject needs to be delivered to the main thread.
+      auto this_shared_ptr = this->shared_from_this();
+      dispatcher_.post([hash_key, this_shared_ptr] { this_shared_ptr->deleteObject(hash_key); });
+    }
+  }
+
+  std::shared_ptr<T> getObject(const T& obj) {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    auto hashed_value = HashFunc{}(obj);
+    auto object_it = object_pool_.find(hashed_value);
+    if (object_it != object_pool_.end()) {
+      auto lock_object = object_it->second.lock();
+      if (lock_object) {
+        return lock_object;
+      }
+    }
+
+    auto this_shared_ptr = this->shared_from_this();
+    std::shared_ptr<T> obj_shared(new T(obj), [hashed_value, this_shared_ptr](T* ptr) {
+      delete ptr;
+      this_shared_ptr->deleteObject(hashed_value);
+    });
+
+    // When inserted, it is possible that the old elements still exist before they can be deleted,
+    // and the insertion will fail and therefore need to be overwritten.
+    auto ret = object_pool_.try_emplace(hashed_value, obj_shared);
+    if (!ret.second) {
+      ASSERT(ret.first->second.use_count() == 0);
+      ret.first->second = obj_shared;
+    }
+    return obj_shared;
+  }
+
+  std::size_t poolSize() const {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    return object_pool_.size();
+  }
+
+  void setThreadIdForTest(const std::thread::id& thread_id) { thread_id_ = thread_id; }
```

Comment:
I've been looking at this for 20 minutes. Although I think this is correct, this is complicated enough that I think we need real tests with threads. Can you please remove this function and actually have the tests use real threads? You can use my ThreadSynchronizer class to relatively easy invoke all of the scenarios that you want to test.

Reply:
The `setThreadIdForTest` function isn't ready to be deleted yet, and I need to use this function to modify the thread id because I'm using Mock's `Dispathcer`, which doesn't help me deliver deleteObject to the corresponding thread. So I still need to switch different thread ids through setThreadIdForTest

Reply:
All changed to real  test with threads,  and remove setThreadIdForTest